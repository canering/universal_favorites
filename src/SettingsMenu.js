/*global chrome browser*/
import React, {useState, useEffect} from "react";
import PropTypes from "prop-types";
import { browserName } from "react-device-detect";
import SyncMenu from "./SyncMenu";
import uuidv1 from "uuid/v1";
import AboutMenu from "./AboutMenu";
import Back from "./images/back.png";

function SettingsMenu(props) {
    const [syncMenu, setSyncMenu] = useState(false);
    const [aboutMenu, setAboutMenu] = useState(false);

    useEffect(() => {
        if(browserName === "Firefox" || browserName === "Edge") {
            browser.storage.local.get(["syncMenu"])
                .then((res) => {
                    if(res.syncMenu) {
                        setSyncMenu(true);
                    }
                }, (err) => console.log(err));
        }
        else {
            chrome.storage.local.get(["syncMenu"], (res) => {
                if(res.syncMenu) {
                    setSyncMenu(true);
                }
            });
        }
    }, []);

    if(syncMenu) {
        return <SyncMenu goBack={closeSyncMenu} sync={props.sync} />;
    }
    if (aboutMenu) {
        return <AboutMenu goBack={closeAboutMenu} />;
    }
    return(
        <div className="menu">
            <button id="onlyBack" className="menuButton" onClick={props.goBack}>
                <img src={Back} alt="back" className="icon" />
            </button>
                
            <div className="settings-menu-item" onClick={openSyncMenu}>Sync With Drive</div>
            <div className="settings-menu-item" onClick={getBookmarks}>Import Bookmarks</div>
            <div className="settings-menu-item" onClick={openAboutMenu}>About</div>
        </div>
    );

    function openAboutMenu() {
        setAboutMenu(true);
    }

    function closeAboutMenu() {
        setAboutMenu(false);
    }

    function openSyncMenu() {
        saveSettingsToBrowser(true);
        setSyncMenu(true);
    }

    function closeSyncMenu() {
        saveSettingsToBrowser(false);
        setSyncMenu(false);
    }

    function saveSettingsToBrowser(state) {
        if(browserName === "Firefox" || browserName === "Edge") {
            browser.storage.local.set({syncMenu:state});
        } else {
            chrome.storage.local.set({syncMenu:state});
        }
    }

    function getBookmarks() {
        if(browserName === "Firefox" || browserName === "Edge") {
            browser.bookmarks.getTree().then((bookmarks) => {
                let bookmarkObject = {type:"folder", label:"imported", data:[], id:uuidv1()};
                if(bookmarks[0]) {
                    printBookmarks(bookmarks[0].children, bookmarkObject);
                }
                props.addImportedFolder(bookmarkObject);
            });
        } else {
            chrome.bookmarks.getTree((bookmarks) => {
                let bookmarkObject = {type:"folder", label:"imported", data:[], id:uuidv1()};
                if(bookmarks[0]) {
                    printBookmarks(bookmarks[0].children, bookmarkObject);
                }
                props.addImportedFolder(bookmarkObject);
            });
        }
    }
      
    function printBookmarks(bookmarks, bookmarkObject) {
        bookmarks.forEach((bookmark) => {
            if (bookmark.children) {
                let subFolder = {type:"folder", label:bookmark.title, data:[], id:uuidv1()};
                printBookmarks(bookmark.children, subFolder);
                bookmarkObject.data.push(subFolder);
            } else {
                bookmarkObject.data.push({type:"link", label:bookmark.title, link:bookmark.url, id:uuidv1()});
            }
        });
    }
}

SettingsMenu.propTypes = {
    sync:PropTypes.func,
    goBack:PropTypes.func,
    addImportedFolder:PropTypes.func
};

export default SettingsMenu;