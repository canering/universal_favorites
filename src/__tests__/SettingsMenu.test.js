import React from "react";
import {shallow, configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import SettingsMenu from "../SettingsMenu";
import chrome from "sinon-chrome";
import toJson from "enzyme-to-json";

window.chrome = chrome;

configure({adapter: new Adapter()});

const props = {
    sync: jest.fn(),
    goBack: jest.fn(),
    addImportedFolder: jest.fn()
};

var component;

beforeEach(() => {
    component = shallow(<SettingsMenu {...props} />);
});

afterEach(() => {
    jest.clearAllMocks();
});

test("settings menu snapshot test", () => {
    let tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test("import bookmarks should create list of links from chrome bookmarks", () => {
    chrome.bookmarks.getTree.yields([{
        children: [
            {title:"folder1", children:[
                {title:"link1", url:"http://example.com"},
                {title:"link2", url:"http://example.com"}
            ]},
            {title:"link3", url:"http://example.com"},
            {title:"link4", url:"http://example.com"}
        ]
    }]);

    component.find(".settings-menu-item").at(1).prop("onClick")({});

    expect(props.addImportedFolder).toHaveBeenCalledTimes(1);
    expect(props.addImportedFolder).toHaveBeenCalledWith(expect.objectContaining({
        type:"folder", label:"imported", id:expect.any(String), data:[
            {type:"folder", label:"folder1", id:expect.any(String), data:[
                {type:"link", label:"link1", id:expect.any(String), link:"http://example.com"},
                {type:"link", label:"link2", id:expect.any(String), link:"http://example.com"}
            ]},
            {type:"link", label:"link3", id:expect.any(String), link:"http://example.com"},
            {type:"link", label:"link4", id:expect.any(String), link:"http://example.com"}
        ]
    }));
});