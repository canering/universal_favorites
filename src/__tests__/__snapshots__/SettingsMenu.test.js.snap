// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings menu snapshot test 1`] = `
<div
  className="menu"
>
  <button
    className="menuButton"
    id="onlyBack"
    onClick={[MockFunction]}
  >
    <img
      alt="back"
      className="icon"
      src="https://img.icons8.com/ios-filled/50/000000/back.png"
    />
  </button>
  <div
    className="settings-menu-item"
    onClick={[Function]}
  >
    Sync Menu
  </div>
  <div
    className="settings-menu-item"
    onClick={[Function]}
  >
    Import Bookmarks
  </div>
</div>
`;
